File: Competetive programming/C++/Unit 2/erdos_number.cpp
#include <iostream>
using namespace std;

int main() {

    int n, m;

    cout << "Enter number of authors: ";
    cin >> n;

    int graph[20][20] = {0};

    cout << "Enter number of collaborations: ";
    cin >> m;

    cout << "\nEnter collaboration pairs:\n";

    for(int i = 0; i < m; i++) {

        int a, b;
        cin >> a >> b;

        graph[a][b] = 1;
        graph[b][a] = 1;
    }

    int erdos[20];
    int queue[20];

    for(int i = 0; i < n; i++)
        erdos[i] = -1;

    int front = 0, rear = 0;

    erdos[0] = 0; // Paul Erdos
    queue[rear++] = 0;

    while(front < rear) {

        int current = queue[front++];

        for(int i = 0; i < n; i++) {

            if(graph[current][i] == 1 && erdos[i] == -1) {

                erdos[i] = erdos[current] + 1;
                queue[rear++] = i;
            }
        }
    }

    cout << "\nErdos Numbers:\n";

    for(int i = 0; i < n; i++)
        cout << "Author " << i << " -> " << erdos[i] << endl;

    return 0;
}

File: Competetive programming/Python/Unit 2/erdos_number.py
print("===== ERDOS NUMBERS =====\n")

authors = {}

n = int(input("Enter number of collaborations: "))

for i in range(n):

    print(f"\nCollaboration {i+1}")

    a = input("Author 1: ")
    b = input("Author 2: ")

    if a not in authors:
        authors[a] = []

    if b not in authors:
        authors[b] = []

    authors[a].append(b)
    authors[b].append(a)

queue = ["Erdos"]
distance = {"Erdos": 0}

while queue:

    current = queue.pop(0)

    for neighbor in authors.get(current, []):

        if neighbor not in distance:

            distance[neighbor] = distance[current] + 1
            queue.append(neighbor)

print("\nErdos Numbers:\n")

for person in distance:
    print(person, "=", distance[person])